lib/charms/vault_k8s/v0/vault_client.py

```diff
@@ -416,3 +447,100 @@ def get_num_raft_peers(self) -> int:
         """Return the number of raft peers."""
         raft_config = self._client.sys.read_raft_config()
         return len(raft_config["data"]["config"]["servers"])
+
+    def _get_autounseal_policy_name(self, relation_id: int) -> str:
+        """Return the policy name for the given relation id."""
+        return f"charm-autounseal-{relation_id}"
+
+    def _get_autounseal_approle_name(self, relation_id: int) -> str:
+        """Return the approle name for the given relation id."""
+        return f"charm-autounseal-{relation_id}"
+
+    def _configure_autounseal_policy(self, mount: str, relation_id: int, key_name: str) -> str:
+        """Create/update a policy within vault to use the transit key."""
+        mount_policy = textwrap.dedent(f"""
+        path "{mount}/encrypt/{key_name}" {{
+            capabilities = ["update"]
+        }}
+
+        path "{mount}/decrypt/{key_name}" {{
+            capabilities = ["update"]
+        }}
+        """)
+        policy_name = self._get_autounseal_policy_name(relation_id)
+        self._client.sys.create_or_update_policy(
+            policy_name,
+            mount_policy,
+        )
+        return policy_name
+
+    def _get_autounseal_key_name(self, relation_id: int) -> str:
+        """Return the key name for the given relation id."""
+        return str(relation_id)
+
+    def _create_autounseal_key(self, mount_point: str, relation_id: int) -> str:
```

Comment:
I don't think we can have charm libraries that depend on other charm libraries, so I'm not sure what your suggesting is possible.

As for separation of concerns, I'm not sure I agree but maybe I could be convinced. In this case I don't want a transit key, I want a key for autounseal functionality. The fact that Vault uses transit to facilitate this is an implementation detail, and irrelevant. It could be something else and it shouldn't affect me.

I think we're coming at the vault client from two different perspectives (and I'm not sure I'm right). I've taken it to be a layer between the Vault API and our charms which abstracts the Vault details away so that we don't need to be concerned about them in the charms (and thus can reduce duplication). If it is solely meant to be a translation layer between the Vault API and our charms, I think we'll see a lot more duplication of code between k8s and machine.

Would love to hear some other opinions on this.

Reply:
I think considering the fact that we can't have libraries talking to each other, it's fine to keep it that way.

As far as the Vault client is concerned, my biggest goal with it is to have a vault client that is idempotent and crash avoidant. This means that vault_client's role is a bit different than the rest, but given the constraints around charm libraries this was something we were willing to go with. I do agree that sharing should be another major goal, but once again I don't see how k8s and machine charms will have their code duplicated in any significant way. Preparing the arguments of a function in 2-3 lines is fine to duplicate in my opinion if it means better separation of concerns and easier maintainability (since we see complex logic inline instead of spread out among files).

although that last point is also highly subjective to me.

Reply:
> my biggest goal with it is to have a vault client that is idempotent and crash avoidant.

I think this is a good point and something we should continue to strive for. I'm going to make a follow-up ticket for this discussion, and I think one way of emphasising this through architecture would be to have two separate classes in the library. One that provides the interface to Vault with all our handling and parsing on top of it (the fault resistant, idempotent interface), and the other that wraps it with our business logic. This is based on some discussion I had with @ghislainbourgeois yesterday, so I don't want to make it sound like it was my idea.

Reply:
Deferring this discussion to a spec: https://warthogs.atlassian.net/browse/TLSENG-269

Reply:
I'm not opposed to that idea at all. also, talked a little about this in my other comment [above](https://github.com/canonical/vault-k8s-operator/pull/365#discussion_r1627961251) just now. feel free to resolve this when you read it.